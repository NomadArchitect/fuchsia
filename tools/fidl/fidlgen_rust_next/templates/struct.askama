{% let is_static = strct.shape.max_out_of_line == 0 %}
{% let name = strct.name.decl_name().camel() %}

{% let params %}
{% if is_static %}
    {% let params = "" %}
{% else %}
    {% let params = "<'buf>" %}
{% endif %}

{% if let Some(doc) = strct.attributes.doc_string() %}
    #[doc = "{{ doc.escape_default() }}"]
{% endif %}
#[derive(
    {% if !strct.is_resource %}
        Clone,
    {% endif %}
    {% if compiler.config.emit_debug_impls %}
        Debug,
    {% endif %}
)]
pub struct {{ name }} {
    {% for member in strct.members %}
        pub {{ member.name.snake() }}: {{ compiler.natural_type(member.ty) }},
    {% endfor %}
}

impl ::fidl_next::Encodable for {{ name }} {
    type Encoded<'buf> = Wire{{ name }}{{ params }};
}

impl<___E> ::fidl_next::Encode<___E> for {{ name }}
where
    {% for member in strct.members %}
        {{ compiler.natural_type(member.ty) }}: ::fidl_next::Encode<___E>,
    {% endfor %}
{
    fn encode(
        &mut self,
        encoder: &mut ___E,
        slot: ::fidl_next::Slot<'_, Self::Encoded<'_>>,
    ) -> Result<(), ::fidl_next::EncodeError> {
        ::fidl_next::munge! {
            let Self::Encoded {
                {% for member in strct.members -%}
                    {{ member.name.snake() }},
                {% endfor %}
            } = slot;
        }

        {% for member in strct.members %}
            ::fidl_next::Encode::encode(&mut self.{{ member.name.snake() }}, encoder, {{ member.name.snake() }})?;
        {% endfor %}
        Ok(())
    }
}

impl ::fidl_next::EncodableOption for Box<{{ name }}> {
    type EncodedOption<'buf> = ::fidl_next::WireBox<'buf, Wire{{ name }}{{ params }}>;
}

impl<___E> ::fidl_next::EncodeOption<___E> for Box<{{ name }}>
where
    ___E: ::fidl_next::Encoder + ?Sized,
    {{ name }}: ::fidl_next::Encode<___E>,
{
    fn encode_option(
        this: Option<&mut Self>,
        encoder: &mut ___E,
        slot: ::fidl_next::Slot<'_, Self::EncodedOption<'_>>,
    ) -> Result<(), ::fidl_next::EncodeError> {
        if let Some(inner) = this {
            ::fidl_next::EncoderExt::encode_next(encoder, inner)?;
            ::fidl_next::WireBox::encode_present(slot);
        } else {
            ::fidl_next::WireBox::encode_absent(slot);
        }

        Ok(())
    }
}

impl{{ params }} ::fidl_next::TakeFrom<Wire{{ name }}{{ params }}> for {{ name }} {
    fn take_from(from: &mut Wire{{ name }}{{ params }}) -> Self {
        Self {
            {% for member in strct.members %}
                {% let member_name = member.name.snake() %}
                {{ member_name }}: ::fidl_next::TakeFrom::take_from(&mut from.{{ member_name }}),
            {% endfor %}
        }
    }
}

/// The wire type corersponding to [`{{ name }}`].
#[derive(
    {% if is_static && !strct.is_resource %}
        Clone,
    {% endif %}
    {% if compiler.config.emit_debug_impls %}
        Debug,
    {% endif %}
)]
#[repr(C)]
pub struct Wire{{ name }}{{ params }} {
    {% for member in strct.members %}
        pub {{ member.name.snake() }}: {{ compiler.wire_type(member.ty) }},
    {% endfor %}
}

unsafe impl<{% if !is_static %}'buf, {% endif %}___D> ::fidl_next::Decode<___D> for Wire{{ name }}{{ params }}
where
    ___D: ?Sized,
    {% for member in strct.members %}
        {{ compiler.wire_type(member.ty) }}: ::fidl_next::Decode<___D>,
    {% endfor %}
{
    fn decode(
        slot: ::fidl_next::Slot<'_, Self>,
        decoder: &mut ___D,
    ) -> Result<(), ::fidl_next::DecodeError> {
        ::fidl_next::munge! {
            let Self {
                {% for member in strct.members -%}
                    mut {{ member.name.snake() }},
                {% endfor %}
            } = slot;
        }

        {% for member in strct.members %}
            {% let member_name = member.name.snake() %}
            ::fidl_next::Decode::decode({{ member_name }}.as_mut(), decoder)?;
            {# type check #}
            {% match member.ty.kind %}
                {% when TypeKind::Array { element_count, element_type } %}
                {% when TypeKind::Vector { element_count, element_type, nullable } %}
                    {% if let Some(limit) = element_count %}
                        let {{ member_name }} = unsafe { {{ member_name }}.deref_unchecked() };
                        {% if nullable %}
                            if let Some({{ member_name }}) = {{ member_name }}.as_ref() {
                        {% endif %}
                                if {{ member_name }}.len() > {{ limit }} {
                                    return Err(::fidl_next::DecodeError::VectorTooLong {
                                        size: {{ member_name }}.len() as u64,
                                        limit: {{ limit }},
                                    });
                                }
                        {% if nullable %}
                            }
                        {% endif %}
                    {% endif %}
                {% when TypeKind::String { element_count, nullable } %}
                    {% if let Some(limit) = element_count %}
                        let {{ member_name }} = unsafe { {{ member_name }}.deref_unchecked() };
                        {% if nullable %}
                            if let Some({{ member_name }}) = {{ member_name }}.as_ref() {
                        {% endif %}
                                if {{ member_name }}.len() > {{ limit }} {
                                    return Err(::fidl_next::DecodeError::VectorTooLong {
                                        size: {{ member_name }}.len() as u64,
                                        limit: {{ limit }},
                                    });
                                }
                        {% if nullable %}
                            }
                        {% endif %}
                    {% endif %}
                {% else %} {# todo #}
            {% endmatch %}
        {% endfor %}
        Ok(())
    }
}