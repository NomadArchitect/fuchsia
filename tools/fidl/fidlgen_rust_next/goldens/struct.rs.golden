// DO NOT EDIT: This file is machine-generated by fidlgen
#![warn(clippy::all)]
#![allow(
    unused_parens,
    unused_variables,
    unused_mut,
    unused_imports,
    unreachable_code,
    nonstandard_style
)]

#[derive(Clone, Debug)]
pub struct BasicStruct {
    pub x: u32,
    pub y: String,
}

impl ::fidl::Encodable for BasicStruct {
    type Encoded<'buf> = WireBasicStruct<'buf>;
}

impl<___E> ::fidl::Encode<___E> for BasicStruct
where
    u32: ::fidl::Encode<___E>,
    String: ::fidl::Encode<___E>,
{
    fn encode(
        &mut self,
        encoder: &mut ___E,
        slot: ::fidl::Slot<'_, Self::Encoded<'_>>,
    ) -> Result<(), ::fidl::EncodeError> {
        ::fidl::munge! {
            let Self::Encoded {
                x,
                y,
            } = slot;
        }
        ::fidl::Encode::encode(&mut self.x, encoder, x)?;
        ::fidl::Encode::encode(&mut self.y, encoder, y)?;
        Ok(())
    }
}

impl ::fidl::EncodableOption for Box<BasicStruct> {
    type EncodedOption<'buf> = ::fidl::WireBox<'buf, WireBasicStruct<'buf>>;
}

impl<___E> ::fidl::EncodeOption<___E> for Box<BasicStruct>
where
    ___E: ::fidl::Encoder + ?Sized,
    BasicStruct: ::fidl::Encode<___E>,
{
    fn encode_option(
        this: Option<&mut Self>,
        encoder: &mut ___E,
        slot: ::fidl::Slot<'_, Self::EncodedOption<'_>>,
    ) -> Result<(), ::fidl::EncodeError> {
        if let Some(inner) = this {
            ::fidl::EncoderExt::encode(encoder, inner)?;
            ::fidl::WireBox::encode_present(slot);
        } else {
            ::fidl::WireBox::encode_absent(slot);
        }

        Ok(())
    }
}

impl<'buf> ::fidl::TakeFrom<WireBasicStruct<'buf>> for BasicStruct {
    fn take_from(from: &mut WireBasicStruct<'buf>) -> Self {
        Self {
            x: ::fidl::TakeFrom::take_from(&mut from.x),
            y: ::fidl::TakeFrom::take_from(&mut from.y),
        }
    }
}

#[derive(Debug)]
#[repr(C)]
pub struct WireBasicStruct<'buf> {
    pub x: ::fidl::u32_le,
    pub y: ::fidl::WireString<'buf>,
}

unsafe impl<'buf, ___D> ::fidl::Decode<___D> for WireBasicStruct<'buf>
where
    ___D: ?Sized,

    ::fidl::u32_le: ::fidl::Decode<___D>,
    ::fidl::WireString<'buf>: ::fidl::Decode<___D>,
{
    fn decode(slot: ::fidl::Slot<'_, Self>, decoder: &mut ___D) -> Result<(), ::fidl::DecodeError> {
        ::fidl::munge! {
            let Self {
                mut x,
                mut y,
            } = slot;
        }
        ::fidl::Decode::decode(x.as_mut(), decoder)?;
        ::fidl::Decode::decode(y.as_mut(), decoder)?;
        Ok(())
    }
}

#[derive(Clone, Debug)]
pub struct Simple {
    pub f1: u8,
    pub f2: bool,
}

impl ::fidl::Encodable for Simple {
    type Encoded<'buf> = WireSimple;
}

impl<___E> ::fidl::Encode<___E> for Simple
where
    u8: ::fidl::Encode<___E>,
    bool: ::fidl::Encode<___E>,
{
    fn encode(
        &mut self,
        encoder: &mut ___E,
        slot: ::fidl::Slot<'_, Self::Encoded<'_>>,
    ) -> Result<(), ::fidl::EncodeError> {
        ::fidl::munge! {
            let Self::Encoded {
                f1,
                f2,
            } = slot;
        }
        ::fidl::Encode::encode(&mut self.f1, encoder, f1)?;
        ::fidl::Encode::encode(&mut self.f2, encoder, f2)?;
        Ok(())
    }
}

impl ::fidl::EncodableOption for Box<Simple> {
    type EncodedOption<'buf> = ::fidl::WireBox<'buf, WireSimple>;
}

impl<___E> ::fidl::EncodeOption<___E> for Box<Simple>
where
    ___E: ::fidl::Encoder + ?Sized,
    Simple: ::fidl::Encode<___E>,
{
    fn encode_option(
        this: Option<&mut Self>,
        encoder: &mut ___E,
        slot: ::fidl::Slot<'_, Self::EncodedOption<'_>>,
    ) -> Result<(), ::fidl::EncodeError> {
        if let Some(inner) = this {
            ::fidl::EncoderExt::encode(encoder, inner)?;
            ::fidl::WireBox::encode_present(slot);
        } else {
            ::fidl::WireBox::encode_absent(slot);
        }

        Ok(())
    }
}

impl ::fidl::TakeFrom<WireSimple> for Simple {
    fn take_from(from: &mut WireSimple) -> Self {
        Self {
            f1: ::fidl::TakeFrom::take_from(&mut from.f1),
            f2: ::fidl::TakeFrom::take_from(&mut from.f2),
        }
    }
}

#[derive(Clone, Debug)]
#[repr(C)]
pub struct WireSimple {
    pub f1: u8,
    pub f2: bool,
}

unsafe impl<___D> ::fidl::Decode<___D> for WireSimple
where
    ___D: ?Sized,

    u8: ::fidl::Decode<___D>,
    bool: ::fidl::Decode<___D>,
{
    fn decode(slot: ::fidl::Slot<'_, Self>, decoder: &mut ___D) -> Result<(), ::fidl::DecodeError> {
        ::fidl::munge! {
            let Self {
                mut f1,
                mut f2,
            } = slot;
        }
        ::fidl::Decode::decode(f1.as_mut(), decoder)?;
        ::fidl::Decode::decode(f2.as_mut(), decoder)?;
        Ok(())
    }
}
